File: src/skillenv/inspect.py
from dataclasses import dataclass

from skillenv.envs import Env
from skillenv.lock import directory_checksum, read_lock
from skillenv.plugins import list_plugins


@dataclass(frozen=True)
class DoctorResult:
    ok: bool
    issues: list[str]


@dataclass(frozen=True)
class DiffResult:
    left_name: str
    right_name: str
    skills_only_left: list[str]
    skills_only_right: list[str]
    plugins_only_left: list[str]
    plugins_only_right: list[str]


def describe_env(env: Env) -> dict:
    lock = read_lock(env)
    return {
        "name": env.name,
        "root": str(env.root),
        "skills": [skill["name"] for skill in lock["skills"]],
        "plugins": list_plugins(env),
    }


def check_env(env: Env) -> DoctorResult:
    issues: list[str] = []
    lock = read_lock(env)
    for required_file in ("config.toml", "skillenv.yml"):
        if not (env.root / required_file).exists():
            issues.append(f"missing {required_file}")

    skills_dir = env.root / "skills"
    if not skills_dir.is_dir():
        issues.append("missing skills directory")
    else:
        for skill_dir in sorted(path for path in skills_dir.iterdir() if path.is_dir()):
            if not (skill_dir / "SKILL.md").is_file():
                issues.append(f"missing SKILL.md: skills/{skill_dir.name}")
    for skill in lock["skills"]:
        checksum = skill.get("checksum")
        if not checksum:
            continue
        skill_dir = env.root / "skills" / skill["name"]
        if skill_dir.is_dir() and directory_checksum(skill_dir) != checksum:
            issues.append(f"checksum mismatch: skills/{skill['name']}")

    return DoctorResult(ok=not issues, issues=issues)


def diff_envs(left: Env, right: Env) -> DiffResult:
    left_summary = describe_env(left)
    right_summary = describe_env(right)
    left_skills = set(left_summary["skills"])
    right_skills = set(right_summary["skills"])
    left_plugins = set(left_summary["plugins"])
    right_plugins = set(right_summary["plugins"])
    return DiffResult(
        left_name=left.name,
        right_name=right.name,
        skills_only_left=sorted(left_skills - right_skills),
        skills_only_right=sorted(right_skills - left_skills),
        plugins_only_left=sorted(left_plugins - right_plugins),
        plugins_only_right=sorted(right_plugins - left_plugins),
    )

File: src/inspect.ts
import { existsSync, readdirSync, readFileSync, statSync } from "node:fs";
import path from "node:path";
import { getAdapter } from "./adapter.js";
import { PROVIDERS } from "./agent/providers.js";
import { directoryChecksum, readLock } from "./lock.js";
import { listPlugins } from "./plugins.js";
import { SKILL_FILE } from "./skill.js";

export interface EnvSummary {
  name: string;
  root: string;
  adapter: string;
  skills: string[];
  plugins: string[];
}

export interface DoctorResult {
  ok: boolean;
  issues: string[];
}

export interface DiffResult {
  leftName: string;
  rightName: string;
  skillsOnlyLeft: string[];
  skillsOnlyRight: string[];
  pluginsOnlyLeft: string[];
  pluginsOnlyRight: string[];
}

export function describeEnv(envRoot: string, name: string): EnvSummary {
  const lock = readLock(envRoot);
  const adapter = readManifestAdapter(envRoot);
  return {
    name,
    root: envRoot,
    adapter,
    skills: lock.skills.map((skill) => skill.name),
    plugins: listPlugins(envRoot),
  };
}

function readManifestAdapter(envRoot: string): string {
  const manifestFile = path.join(envRoot, "skillenv.yml");
  try {
    const text = readFileSync(manifestFile, "utf8");
    const match = /^adapter:\s*(\S+)/m.exec(text);
    return match?.[1] ?? "codex";
  } catch {
    return "codex";
  }
}

/**
 * Health check: manifest/config presence, adapter-specific required files,
 * SKILL.md in every skill directory, and lock checksum verification.
 */
export interface ProviderReadiness {
  id: string;
  displayName: string;
  ready: boolean;
  missing: string;
}

/** Which agent providers have their credentials configured (never fatal). */
export function providerReadiness(): ProviderReadiness[] {
  return Object.values(PROVIDERS).map((preset) => {
    if (!preset.apiKeyEnv) {
      return { id: preset.id, displayName: preset.displayName, ready: true, missing: "" };
    }
    const hasKey = Boolean(process.env[preset.apiKeyEnv]);
    const needsBase = preset.baseUrlEnv ? Boolean(process.env[preset.baseUrlEnv]) : true;
    if (preset.baseUrlEnv && !preset.baseUrl && !needsBase) {
      return {
        id: preset.id,
        displayName: preset.displayName,
        ready: false,
        missing: `${preset.apiKeyEnv} and ${preset.baseUrlEnv}`,
      };
    }
    return {
      id: preset.id,
      displayName: preset.displayName,
      ready: hasKey,
      missing: hasKey ? "" : preset.apiKeyEnv,
    };
  });
}

export function checkEnv(envRoot: string, name: string): DoctorResult {
  const issues: string[] = [];
  const adapterId = readManifestAdapter(envRoot);

  let adapter;
  try {
    adapter = getAdapter(adapterId);
  } catch (error) {
    issues.push((error as Error).message);
    adapter = null;
  }

  const requiredFiles = ["skillenv.yml", ...(adapter ? adapter.requiredFiles : ["config.toml"])];
  for (const file of requiredFiles) {
    if (!existsSync(path.join(envRoot, file))) {
      issues.push(`missing ${file}`);
    }
  }

  const skillsDir = path.join(envRoot, "skills");
  if (!statSync(skillsDir, { throwIfNoEntry: false })?.isDirectory()) {
    issues.push("missing skills directory");
  } else {
    for (const entry of readdirSync(skillsDir).sort()) {
      const skillDir = path.join(skillsDir, entry);
      if (statSync(skillDir, { throwIfNoEntry: false })?.isDirectory() && !existsSync(path.join(skillDir, SKILL_FILE))) {
        issues.push(`missing ${SKILL_FILE}: skills/${entry}`);
      }
    }
  }

  const lock = readLock(envRoot);
  for (const skill of lock.skills) {
    if (!skill.checksum) continue;
    const skillDir = path.join(envRoot, "skills", skill.name);
    if (statSync(skillDir, { throwIfNoEntry: false })?.isDirectory() && directoryChecksum(skillDir) !== skill.checksum) {
      issues.push(`checksum mismatch: skills/${skill.name}`);
    }
  }

  return { ok: issues.length === 0, issues };
}

export function diffEnvs(leftRoot: string, leftName: string, rightRoot: string, rightName: string): DiffResult {
  const left = describeEnv(leftRoot, leftName);
  const right = describeEnv(rightRoot, rightName);
  const setDifference = (a: string[], b: string[]): string[] =>
    a.filter((item) => !b.includes(item)).sort();
  return {
    leftName,
    rightName,
    skillsOnlyLeft: setDifference(left.skills, right.skills),
    skillsOnlyRight: setDifference(right.skills, left.skills),
    pluginsOnlyLeft: setDifference(left.plugins, right.plugins),
    pluginsOnlyRight: setDifference(right.plugins, left.plugins),
  };
}
